docker/docker.go

```diff
@@ -22,201 +22,460 @@ limitations under the License.
-		if ns[3].Value == "*" {
-			// example ns: /intel/linux/docker/*/cpu_stats/throttling_data/priods
-			for _, container := range d.containersInfo {
-				// copy namespace so it doesn't get altered each time
-				nscopy := make(core.Namespace, len(ns))
-				copy(nscopy, ns)
-
-				dockerID := container.Id
-				// get short version of container ID
-				nscopy[3].Value = dockerID[:12]
-
-				if err := d.getStats(dockerID); err != nil {
-					return nil, err
-				}
+			// remove added static element (`value`)
+			if metricName[len(metricName)-1] == "value" {
+				metricName = metricName[:len(metricName)-1]
+			}
 
-				// only "cgroup.Stats" part of namespace is sent to retrieve value (cpu_stats/throttling_data/periods)
-				mt := plugin.MetricType{
-					Data_:      d.tools.GetValueByNamespace(d.stats, nscopy.Strings()[4:]),
+			if !isDynamic {
+				metric := plugin.MetricType{
 					Timestamp_: time.Now(),
-					Namespace_: nscopy,
-					Version_:   VERSION,
+					Namespace_: ns,
+					Data_:      utils.GetValueByNamespace(d.containers[id], metricName),
+					Tags_:      mt.Tags(),
+					Config_:    mt.Config(),
 				}
-				metrics = append(metrics, mt)
-			}
-		} else {
-			// example ns: /intel/linux/docker/31068893a2bc/cpu_stats/throttling_data/priods
 
-			// extracted docker id from namespace is extended to long one
-			dockerID, err := d.extendDockerID(ns[3].Value)
-			if err != nil {
-				return nil, err
-			}
+				metrics = append(metrics, metric)
+				continue
 
-			// long id is required to get stats for docker
-			if err := d.getStats(dockerID); err != nil {
-				return nil, err
 			}
 
-			// only "cgroup.Stats" part of namespace is sent to retrieve value (cpu_stats/throttling_data/periods)
-			mt := plugin.MetricType{
-				Data_:      d.tools.GetValueByNamespace(d.stats, ns.Strings()[4:]),
-				Timestamp_: time.Now(),
-				Namespace_: ns,
-				Version_:   VERSION,
-			}
-			metrics = append(metrics, mt)
-		}
+			// take the element of metricName which precedes the first dynamic element
+			// e.g. {"filesystem", "*", "usage"}
+			// 	-> statsType will be "filesystem",
+			// 	-> scope of metricName will be decreased to {"*", "usage"}
+
+			indexOfDynamicElement := indexes[0]
+			statsType := metricName[indexOfDynamicElement-1]
+			metricName = metricName[indexOfDynamicElement:]
+
+			switch statsType {
+			case "filesystem":
+				// get docker filesystem statistics
+				devices := []string{}
+
+				if metricName[0] == "*" {
+					// when device name is requested as as asterisk - take all available filesystem devices
+					for deviceName := range d.containers[id].Stats.Filesystem {
+						devices = append(devices, deviceName)
+					}
+				} else {
+					// device name is requested explicitly
+					devices = append(devices, metricName[0])
+				}
+
+				for _, device := range devices {
+					rns := make([]core.NamespaceElement, len(ns))
+					copy(rns, ns)
+
+					rns[indexOfDynamicElement+3].Value = device
```

Comment:
It seems that here instead of "3" you can use const `lenghtOfNsPrefix` as it is in [this](https://github.com/intelsdi-x/snap-plugin-collector-docker/pull/32/files#diff-b715a530d535012e418a564a3eafedb9R238) line.